FWCore/Framework/src/EventSetupRecordIOVQueue.cc

```diff
@@ -0,0 +1,132 @@
+      waitForIOVsInFlight_->wait_for_all();
+    }
+
+    void EventSetupRecordIOVQueue::setNewIntervalForAnySubProcess() {
+      bool newIntervalForAnySubProcess = false;
+      for (auto const& recordProvider : recordProviders_) {
+        if (recordProvider->intervalStatus() == EventSetupRecordProvider::IntervalStatus::NewInterval) {
+          newIntervalForAnySubProcess = true;
+          break;
+        }
+      }
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->setNewIntervalForAnySubProcess(newIntervalForAnySubProcess);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::checkForNewIOVs(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                   WaitingTaskList& endIOVWaitingTasks,
+                                                   bool newEventSetupImpl) {
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->newIntervalForAnySubProcess()) {
+          startNewIOVAsync(taskToStartAfterIOVInit, endIOVWaitingTasks);
+          return;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->intervalStatus() != EventSetupRecordProvider::IntervalStatus::Invalid) {
+          endIOVWaitingTasks.add(endIOVWaitingTask_);
+          break;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->continueIOV(newEventSetupImpl);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::startNewIOVAsync(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                    WaitingTaskList& endIOVWaitingTasks) {
+      ++cacheIdentifier_;
+      {
+        // Let the old IOV end when all the lumis using it are done.
+        // otherwise we'll deadlock when there is only one thread.
+        WaitingTaskHolder tmp{std::move(endIOVTaskHolder_)};
+      }
+      auto taskHolder = std::make_shared<WaitingTaskHolder>(taskToStartAfterIOVInit);
+      auto startIOVForRecord =
+          [this, taskHolder, &endIOVWaitingTasks](edm::LimitedTaskQueue::Resumer iResumer) mutable {
+            try {
+              unsigned int iovIndex = 0;
+              for (; iovIndex < nConcurrentIOVs_; ++iovIndex) {
+                bool expected = true;
+                if (isAvailable_[iovIndex]->compare_exchange_strong(expected, false)) {
+                  break;
+                }
+              }
+              // Should never fail, just a sanity check
+              if (iovIndex == nConcurrentIOVs_) {
+                throw edm::Exception(edm::errors::LogicError)
+                    << "EventSetupRecordIOVQueue::startNewIOVAsync\n"
+                    << "Couldn't find available IOV slot. This should never happen.\n"
+                    << "Contact a Framework Developer\n";
+              }
+              for (auto recordProvider : recordProviders_) {
+                recordProvider->initializeForNewIOV(iovIndex, cacheIdentifier_);
```

Comment:
cacheIdentifier_ could be read while it is being updated.

Reply:
See answer to previous comment.